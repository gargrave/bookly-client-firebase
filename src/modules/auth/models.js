// @flow
import type {
  LoginErrors,
  LoginUser,
  PasswordReset,
  PasswordResetErrors,
  RegisterErrors,
  RegisterUser,
  User,
  UserWithMetadata,
} from './flowtypes';

export const userModel = {
  empty(): User {
    return {
      displayName: '',
      email: '',
      emailVerified: false,
      id: '',
      lastLogin: '',
      registered: '',
    };
  },

  fromAPI(user: UserWithMetadata): User {
    const {
      displayName,
      email,
      emailVerified,
      uid,
    } = user;
    const {
      creationTime,
      lastSignInTime,
    } = user.metadata;

    return {
      displayName,
      email,
      emailVerified,
      id: uid,
      lastLogin: lastSignInTime,
      registered: creationTime,
    };
  },
};

export const loginUserModel = {
  empty(): LoginUser {
    return {
      email: '',
      password: '',
    };
  },

  emptyErrors(): LoginErrors {
    return {
      email: '',
      password: '',
    };
  },

  toAPI(data: LoginUser): LoginUser {
    let payload: LoginUser = {
      email: data.email,
      password: data.password,
    };
    return payload;
  },
};

export const registerUserModel = {
  empty(): RegisterUser {
    return {
      email: '',
      password: '',
      passwordConfirm: '',
    };
  },

  emptyErrors(): RegisterErrors {
    return {
      email: '',
      password: '',
      passwordConfirm: '',
    };
  },

  toAPI(data: RegisterUser): RegisterUser {
    return {
      email: data.email,
      password: data.password,
      passwordConfirm: data.passwordConfirm,
    };
  },
};

export const passwordResetModel = {
  empty(): PasswordReset {
    return {
      email: '',
    };
  },

  emptyErrors(): PasswordResetErrors {
    return {
      email: '',
    };
  },
};
