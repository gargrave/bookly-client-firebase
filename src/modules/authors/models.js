// @flow
import type { Author, AuthorErrors } from './flowtypes';
import type { FbDoc } from '../core/flowtypes';

export const authorModel = {
  empty(): Author {
    return {
      firstName: '',
      lastName: '',
    };
  },

  emptyErrors(): AuthorErrors {
    return {
      firstName: '',
      lastName: '',
    };
  },

  editable(author: Author): Author {
    return {
      created: author.created,
      firstName: author.firstName,
      id: author.id,
      lastName: author.lastName,
    };
  },

  toAPI(data: Author): Author {
    let payload: any = {
      firstName: data.firstName.trim() || '',
      lastName: data.lastName.trim() || '',
    };

    ['id', 'created'].forEach((val) => {
      if (data[val]) {
        payload[val] = data[val];
      }
    });

    return payload;
  },

  fromAPI(doc: FbDoc): Author {
    const {
      firstName,
      lastName,
      created,
      updated,
    } = doc.data();

    return {
      id: doc.id,
      firstName,
      lastName,
      created,
      updated,
    };
  },
};
