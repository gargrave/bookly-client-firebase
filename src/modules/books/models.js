// @flow
import type { Author } from '../authors/flowtypes';
import type { FbDoc } from '../core/flowtypes';
import type { Book, BookErrors } from './flowtypes';

function hydrateAuthor(authors, id): Author {
  const author = authors.find((a) => a.id === id);
  return author || {
    firstName: '',
    lastName: '',
  };
}

export function refreshBookAuthor(book: Book, authors: Author[]) {
  book.author = hydrateAuthor(authors, book.author.id || book.authorId);
}

export const bookModel = {
  empty(): Book {
    return {
      author: {
        id: '',
        firstName: '',
        lastName: '',
      },
      title: '',
    };
  },

  emptyErrors(): BookErrors {
    return {
      title: '',
      author: '',
    };
  },

  editable(book: Book): Book {
    return {
      author: book.author,
      created: book.created,
      id: book.id,
      title: book.title.trim(),
    };
  },

  toAPI(data: Book): Book {
    let payload: any = {
      title: data.title.trim(),
      authorId: data.author.id,
    };

    ['id', 'created'].forEach((val) => {
      if (data[val]) {
        payload[val] = data[val];
      }
    });

    return payload;
  },

  fromAPI(doc: FbDoc, authors: Author[]): Book {
    const {
      title,
      authorId,
      created,
      updated,
    } = doc.data();

    return {
      id: doc.id,
      title,
      author: hydrateAuthor(authors, authorId),
      created,
      updated,
    };
  },
};
