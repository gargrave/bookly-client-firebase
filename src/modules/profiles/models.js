// @flow
import type { FbDoc } from '../core/flowtypes';
import type { Profile, ProfileErrors } from './flowtypes';

export const profileModel = {
  empty(): Profile {
    return {
      firstName: '',
      lastName: '',
    };
  },

  emptyErrors(): ProfileErrors {
    return {
      firstName: '',
      lastName: '',
    };
  },

  editable(profile: Profile): Profile {
    return {
      created: profile.created,
      firstName: profile.firstName,
      id: profile.id,
      lastName: profile.lastName,
    };
  },

  toAPI(data: Profile): Profile {
    let payload: any = {
      firstName: data.firstName.trim() || '',
      lastName: data.lastName.trim() || '',
    };

    ['id', 'created'].forEach((val) => {
      if (data[val]) {
        payload[val] = data[val];
      }
    });

    return payload;
  },

  fromAPI(doc: FbDoc): Profile {
    const {
      created,
      firstName,
      lastName,
      updated,
    } = doc.data();

    return {
      created,
      firstName,
      id: doc.id,
      lastName,
      updated,
    };
  },
};
