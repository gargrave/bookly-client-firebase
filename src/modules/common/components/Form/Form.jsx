// @flow
import React from 'react'
import { any, array, func, bool, string } from 'prop-types'

import { buildClass } from '../../../../utils/cssHelpers'

import Alert from '../Alert/Alert'
import Button from '../Button'
import ButtonRow from '../ButtonRow/ButtonRow'

import styles from './Form.css'

type Props = {
  cancelBtnText?: string,
  children?: any,
  classes?: string[],
  disabled?: boolean,
  onCancel?: Function,
  onSubmit: Function,
  submitBtnText?: string,
  submitDisabled?: boolean,
  topLevelError?: string,
}

const Form = ({
  cancelBtnText,
  children,
  classes = [],
  disabled,
  onCancel,
  onSubmit,
  submitBtnText,
  submitDisabled = false,
  topLevelError,
}: Props) => {
  return (
    <div>
      {topLevelError && <Alert message={topLevelError} type="danger" />}
      <form
        className={buildClass(styles.form, ...(classes || []))}
        onSubmit={onSubmit}
        noValidate
      >
        {children}

        <ButtonRow>
          <Button
            canSubmit={true}
            disabled={submitDisabled || disabled || false}
            onClick={onSubmit}
            position="left"
            text={submitBtnText || 'Submit'}
            type="success"
          />

          {onCancel && (
            <Button
              onClick={onCancel}
              position="right"
              text={cancelBtnText || 'Cancel'}
              type="light"
            />
          )}
        </ButtonRow>
      </form>
    </div>
  )
}

Form.propTypes = {
  cancelBtnText: string,
  children: any,
  classes: array,
  disabled: bool,
  onCancel: func,
  onSubmit: func.isRequired,
  submitBtnText: string,
  submitDisabled: bool,
  topLevelError: string,
}

export default Form
