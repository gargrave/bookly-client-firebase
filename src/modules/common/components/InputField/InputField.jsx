// @flow
import React from 'react'
import { bool, func, number, oneOf, string } from 'prop-types'
import styled from 'react-emotion'

import { colors } from '../../../../styles'
import { clamp } from '../../../../utils/mathHelpers'

type Props = {
  boundValue: string,
  disabled?: boolean,
  error?: string,
  label?: string,
  maxLength?: number,
  name: string,
  onInputChange: Function,
  placeholder?: string,
  type?: string,
}

const acceptableTypes = ['email', 'password', 'search', 'text']

const StyledInput = styled('input')`
  border-color: ${props => props.invalid && colors.error} !important;
`

const StyledError = styled('p')`
  color: ${colors.error};
`

const InputField = ({
  boundValue,
  disabled,
  error,
  label,
  maxLength = 255,
  name,
  onInputChange,
  placeholder,
  type,
}: Props) => {
  return (
    <div className="input-field">
      {label && <label htmlFor={name}>{label}:</label>}

      <StyledInput
        disabled={disabled || false}
        id={name}
        invalid={!!error}
        maxLength={clamp(maxLength, 1, 255)}
        name={name}
        onChange={onInputChange}
        placeholder={placeholder || ''}
        type={type || 'text'}
        value={boundValue}
      />

      {error && <StyledError className="error">{error}</StyledError>}
    </div>
  )
}

InputField.propTypes = {
  boundValue: string.isRequired,
  disabled: bool,
  error: string,
  maxLength: number,
  label: string,
  name: string.isRequired,
  onInputChange: func.isRequired,
  placeholder: string,
  type: oneOf(acceptableTypes),
}

export default InputField
