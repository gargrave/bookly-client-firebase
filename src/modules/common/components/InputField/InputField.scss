@import '../../../../styles/_colors';

.error {
  color: $errorColor;
}

input.invalid {
  border-color: $errorColor;
}
