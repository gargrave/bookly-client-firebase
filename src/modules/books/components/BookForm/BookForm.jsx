// @flow
import React from 'react'
import { array, bool, func, object, shape, string } from 'prop-types'

import type { Author } from '../../../authors/flowtypes'
import type { Book, BookErrors } from '../../flowtypes'

import AuthorSelect from '../../../authors/components/AuthorSelect'
import Form from '../../../common/components/Form/Form'
import InputField from '../../../common/components/InputField/InputField'

type Props = {
  authors: Author[],
  book: Book,
  disabled?: boolean,
  errors: BookErrors,
  onAuthorChange: Function,
  onCancel: Function,
  onInputChange: Function,
  onSubmit: Function,
  preselectedAuthor?: Author,
  submitDisabled?: boolean,
  topLevelError?: string,
}

const BookForm = ({
  authors,
  book,
  disabled,
  errors,
  onAuthorChange,
  onCancel,
  onInputChange,
  onSubmit,
  preselectedAuthor,
  submitDisabled = false,
  topLevelError,
}: Props) => {
  const { author, sortBy, title } = book

  return (
    <Form
      classes={['book-form']}
      disabled={disabled}
      onCancel={onCancel}
      onSubmit={onSubmit}
      submitDisabled={submitDisabled}
      topLevelError={topLevelError}
    >
      <InputField
        boundValue={title}
        disabled={disabled || false}
        error={errors.title}
        label="Title"
        name="title"
        onInputChange={onInputChange}
      />

      <InputField
        boundValue={sortBy}
        disabled={disabled || false}
        error={errors.sortBy}
        label="Sort By"
        name="sortBy"
        onInputChange={onInputChange}
        placeholder={title}
      />

      <AuthorSelect
        author={preselectedAuthor || author}
        authors={authors}
        disabled={disabled || false}
        error={errors.author}
        onChange={onAuthorChange}
      />
    </Form>
  )
}

BookForm.propTypes = {
  authors: array.isRequired,
  book: shape({
    title: string.isRequired,
    author: object.isRequired,
  }).isRequired,
  disabled: bool,
  errors: shape({
    title: string.isRequired,
    author: string.isRequired,
  }).isRequired,
  onAuthorChange: func.isRequired,
  onCancel: func.isRequired,
  onInputChange: func.isRequired,
  onSubmit: func.isRequired,
  preselectedAuthor: object,
  submitDisabled: bool,
  topLevelError: string,
}

export default BookForm
