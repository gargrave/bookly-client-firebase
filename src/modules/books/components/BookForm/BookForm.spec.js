import React from 'react';
import { shallow } from 'enzyme';

import { authorModel } from '../../../authors/models';
import { bookModel } from '../../models';

import Form from '../../../common/components/Form/Form';
import InputField from '../../../common/components/InputField/InputField';

import BookForm from './BookForm';

describe('BookForm', () => {
  let props;
  let component;

  describe('with "book" populated', () => {
    beforeEach(() => {
      props = {
        authors: [],
        book: {
          title: '',
          author: {
            id: '98hfjsdbf9s8fhdf',
            ...authorModel.empty(),
          },
        },
        disabled: false,
        errors: bookModel.emptyErrors(),
        onAuthorChange: jest.fn(),
        onCancel: jest.fn(),
        onInputChange: jest.fn(),
        onSubmit: jest.fn(),
        topLevelError: '',
      };

      component = shallow(<BookForm {...props} />);
    });

    test('renders correctly', () => {
      expect(component).toMatchSnapshot();
      expect(component.find(Form)).toHaveLength(1);
      expect(component.find(InputField)).toHaveLength(1);
    });
  });
});
