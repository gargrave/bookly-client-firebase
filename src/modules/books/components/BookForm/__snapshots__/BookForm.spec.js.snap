// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookForm with "book" populated renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BookForm
    authors={Array []}
    book={
      Object {
        "author": Object {
          "firstName": "",
          "id": "98hfjsdbf9s8fhdf",
          "lastName": "",
        },
        "sortBy": "",
        "title": "",
      }
    }
    disabled={false}
    errors={
      Object {
        "author": "",
        "sortBy": "",
        "title": "",
      }
    }
    onAuthorChange={[Function]}
    onCancel={[Function]}
    onInputChange={[Function]}
    onSubmit={[Function]}
    topLevelError=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <InputField
          boundValue=""
          disabled={false}
          error=""
          label="Title"
          name="title"
          onInputChange={[Function]}
        />,
        <InputField
          boundValue=""
          disabled={false}
          error=""
          label="Sort By"
          name="sortBy"
          onInputChange={[Function]}
        />,
        <AuthorSelect
          author={
            Object {
              "firstName": "",
              "id": "98hfjsdbf9s8fhdf",
              "lastName": "",
            }
          }
          authors={Array []}
          disabled={false}
          error=""
          onChange={[Function]}
        />,
      ],
      "classes": Array [
        "book-form",
      ],
      "disabled": false,
      "onCancel": [Function],
      "onSubmit": [Function],
      "submitDisabled": false,
      "topLevelError": "",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "boundValue": "",
          "disabled": false,
          "error": "",
          "label": "Title",
          "name": "title",
          "onInputChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "boundValue": "",
          "disabled": false,
          "error": "",
          "label": "Sort By",
          "name": "sortBy",
          "onInputChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "author": Object {
            "firstName": "",
            "id": "98hfjsdbf9s8fhdf",
            "lastName": "",
          },
          "authors": Array [],
          "disabled": false,
          "error": "",
          "onChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <InputField
            boundValue=""
            disabled={false}
            error=""
            label="Title"
            name="title"
            onInputChange={[Function]}
          />,
          <InputField
            boundValue=""
            disabled={false}
            error=""
            label="Sort By"
            name="sortBy"
            onInputChange={[Function]}
          />,
          <AuthorSelect
            author={
              Object {
                "firstName": "",
                "id": "98hfjsdbf9s8fhdf",
                "lastName": "",
              }
            }
            authors={Array []}
            disabled={false}
            error=""
            onChange={[Function]}
          />,
        ],
        "classes": Array [
          "book-form",
        ],
        "disabled": false,
        "onCancel": [Function],
        "onSubmit": [Function],
        "submitDisabled": false,
        "topLevelError": "",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "boundValue": "",
            "disabled": false,
            "error": "",
            "label": "Title",
            "name": "title",
            "onInputChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "boundValue": "",
            "disabled": false,
            "error": "",
            "label": "Sort By",
            "name": "sortBy",
            "onInputChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "author": Object {
              "firstName": "",
              "id": "98hfjsdbf9s8fhdf",
              "lastName": "",
            },
            "authors": Array [],
            "disabled": false,
            "error": "",
            "onChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
